File: src/scala/axis-cli/src/main/scala/command/Item.scala
package command

import data.{AxisItem, DataUtils}
import picocli.CommandLine.{Command, HelpCommand, Option, Parameters}

import java.time.OffsetDateTime

@Command(name="item", subcommands = Array(classOf[HelpCommand]))
class Item {
    @Command
    def create(@Parameters title: String,
               @Option(names=Array("--universe","-u"), required = true) universeName: String,
               @Option(names=Array("--description","-d"), defaultValue = "") description: String,
               @Option(names=Array("--priority","-p"), defaultValue = "0") priority: Int,
                @Option(names=Array("--due")) timeDue: OffsetDateTime): Unit = {
        AxisItem.create(title, universeName, description, timeDue: OffsetDateTime, priority: Int)
    }

    @Command
    def list(): Unit = {
        var items = AxisItem.list()
        if (items.nonEmpty) {
            DataUtils.printNonEmptyItems(items.collect(r => r.toMap()), AxisItem.printFields)
        }
    }

    @Command
    def search(@Parameters keyword: String, @Option(names=Array("--search-description","-d"), defaultValue = "false") searchDescription: Boolean): Unit = {
        var items = AxisItem.search(keyword, searchDescription)
        if (items.nonEmpty) {
            DataUtils.printNonEmptyItems(items.collect(r => r.toMap()), AxisItem.printFields)
        }
    }

    @Command
    def show(@Parameters id: Int): Unit = {
        var items = AxisItem.find(id)
        DataUtils.printNonEmptyItems(items.collect(r => r.toMap()), AxisItem.printFields)
    }

    @Command
    def delete(@Parameters id: Int): Unit = {
        AxisItem.delete(id)
    }

    @Command
    def update(@Parameters id: Int, @Option(names = Array("--title", "-t")) newTitle: String,
               @Option(names = Array("--description", "-d")) newDescription: String,
               @Option(names = Array("--priority", "-p"), defaultValue = "0") newPriority: Int,
               @Option(names = Array("--due")) newTimeDue: OffsetDateTime,
               @Option(names = Array("--status", "-s")) newStatus: String): Unit = {
        AxisItem.update(id, newTitle, newDescription, newTimeDue, newPriority, newStatus)
    }

    @Command
    def move(@Parameters id: Int, @Option(names=Array("--universe", "-u"), required = true) universeName: String): Unit = {
        AxisItem.move(id, universeName)
    }
}

File: src/swift/axis-cli/Sources/axis-cli/command/Item.swift
import ArgumentParser
import Foundation


struct Item: ParsableCommand {
    static var configuration = CommandConfiguration(
            subcommands: [Create.self, List.self, Search.self, Show.self, Delete.self, Update.self, Move.self])
}


extension Item {
    struct Create: ParsableCommand {
        @Argument var title: String
        @Option(name: .shortAndLong) var description: String?
        @Option(name: .shortAndLong) var universe: String
        @Option(name: .long) var due: String?
        @Option(name: .shortAndLong) var priority: UInt?

        mutating func run() throws {
            let formatter = ISO8601DateFormatter()
            // formatter.timeZone = TimeZone.current
            let timeDue = due != nil ? formatter.date(from: due!)! : nil
            try AxisItem.Create(title: title, description: description, universeName: universe, timeDue: timeDue, priority: priority)
        }
    }

    struct List: ParsableCommand {
        mutating func run() throws {
            let items = try AxisItem.List()
            AxisItem.printRecords(items: items)
        }
    }

    struct Search: ParsableCommand {
        @Argument var keyword: String
        @Flag(name: [.long, .customShort("d")]) var searchDescription = false

        mutating func run() throws {
            let items = try AxisItem.Search(keyword: keyword, searchDescription: searchDescription)
            AxisItem.printRecords(items: items)
        }
    }

    struct Show: ParsableCommand {
        @Argument var id: Int?

        mutating func run() throws {
            let item = try AxisItem.Find(id: id!)
            AxisItem.printRecords(items: [item])
        }
    }

    struct Delete: ParsableCommand {
        @Argument var id: Int?

        mutating func run() throws {
            try AxisItem.Delete(id: id!)
        }
    }

    struct Update: ParsableCommand {
        @Argument var id: Int?
        @Option(name: .shortAndLong) var title: String?
        @Option(name: .shortAndLong) var description: String?
        @Option(name: .long) var due: String?
        @Option(name: .shortAndLong) var priority: UInt?

        mutating func run() throws {
            let formatter = ISO8601DateFormatter()
            // formatter.timeZone = TimeZone.current
            let timeDue = due != nil ? formatter.date(from: due!)! : nil
            try AxisItem.Update(id: id!, title: title, description: description, timeDue: timeDue, priority: priority)
        }
    }

    struct Move: ParsableCommand {
        @Argument var id: Int?
        @Option(name: .shortAndLong) var universe: String

        mutating func run() throws {
            try AxisItem.Update(id: id!, universeName: universe)
        }
    }
}